Assets/Scripts/BoardManager.cs

```diff
@@ -22,47 +44,168 @@ public class BoardManager : MonoBehaviour
     private Player player;
     private Transform boardHolder;
     private List<Vector3> gridPositions = new List<Vector3>();
+    private List<Vector3> outerWallGridPositions = new List<Vector3>();
+    private List<Vector3> innerWallGridPositions = new List<Vector3>();
 
-    void InitialiseList()
+    private int numberOfRoom = 1;
+
+    void BoardSetup()
     {
-        gridPositions.Clear();
+        columnsCount = Random.Range(columns.minimum, columns.maximum + 1);
+        rowsCount = Random.Range(rows.minimum, rows.maximum + 1);
+
+        numberOfRoom = Random.Range(numberOfRoomParts.minimum, numberOfRoomParts.maximum + 1) - 1;
+
+        boardHolder = new GameObject("Board").transform;
+
+        int x = 0;
+        int y = 0;
+        List<string> roomPosition = new List<string>();
+        List<Vector2> m = new List<Vector2>() { new Vector2(x, y) };
+        List<string> validPosition = new List<string>() { "top", "right", "bottom" };
```

Comment:
1. `validPosition` is a list, containing multiple values -> It must be pural `validPositions`

2. You can make an enum instead of using strings which are prone to typo errors, requires allocations and their comparisons are costly.

3. I suppose you didn't include "left" option in this first list to have a generic `Random.Range(0, 3)`. If it is the case, add it back and simply use `Random.Range(0, validPosition.Length)`

4. I assume you fixed the max numberOfRoomPart to 4 by default because higher it breaks your algorithm. Either modify your algorithm to allow making more than 4 room parts without having 2 on the same position or just add some documentation saying it. 
It can be modified easily by simply adding the room positions (x, y) in a list and checking if the new position is available in the list, if no remove the direction from the validPosition list and try another direction.

5. Re allocating a new list each time isn't very efficient. It is acceptable because this is not sensitive code (meaning it will not be executed often) but rethinking the algorithm can make it less memory costly and more generic.